server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -818,6 +818,7 @@ private void executePipelines(
 
         // reset the reroute flag, at the start of a new pipeline execution this document hasn't been rerouted yet
         ingestDocument.resetReroute();
+        indexRequest.addPipeline(pipelineId);
```

Comment:
Notes from conversation elsewhere: 

1.) Shuffle the code around so that this is `ingestDocument.addPipeline(pipelineId);` instead.
2.) a.) Handle the unhappy path that the if/throw covers below -- probably by moving this line to after the if/throw.
2.) b.) Or you could move the line into `executePipeline` itself in which case that would be more automatic...

Reply:
I've done 2a -- 1 is turning out to be trickier than you'd think.